src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+                    y_pos = self.rz_y_lb + self._screw_height / 2.0
+                    screw_cluster_xcenter = rng.choice(screw_cluster_xcenters)
+                    x_pos = rng.normal(screw_cluster_xcenter)
+                    # if this position doesn't make this screw in collision
+                    # with another screw, then put the screw here.
+                    if self._surface_xy_is_valid(x_pos, y_pos, existing_xys):
+                        screw_name_to_pos[f"screw_{si}"] = (x_pos, y_pos)
+                        break
+            init_state, goal_atoms = \
+                self._get_init_state_and_goal_atoms_from_positions(
+                screw_name_to_pos, rng)
+            tasks.append(Task(init_state, goal_atoms))
+        return tasks
+
+    def _get_init_state_and_goal_atoms_from_positions(
+            self, screw_name_to_pos: Dict[str, Tuple[float, float]],
+            rng: np.random.Generator) -> Tuple[State, Set[GroundAtom]]:
+        data: Dict[Object, Array] = {}
+        goal_atoms: Set[GroundAtom] = set()
+        # Select a random screw that needs to be in the receptacle as the goal
+        # GroundAtom.
+        goal_screw_name = rng.choice(list(screw_name_to_pos.keys()))
+
+        # Create screw objects.
+        for screw_name, (x_pos, y_pos) in screw_name_to_pos.items():
+            screw_obj = Object(screw_name, self._screw_type)
+            data[screw_obj] = np.array(
+                [x_pos, y_pos, self._screw_width, self._screw_height, 0.0])
+            # If the screw is the goal screw, then add it being in the
+            # receptacle to the goal atoms.
+            if screw_name == goal_screw_name:
+                goal_atoms.add(
+                    GroundAtom(self._ScrewInReceptacle,
+                               [screw_obj, self._receptacle]))
+
+        # Create receptacle object such that it is attached to the
+        # right wall and halfway between the ceiling and floor.
+        data[self._receptacle] = np.array([
+            self.rz_x_ub - (self._receptacle_width / 2.0),
+            (self.rz_y_ub - self.rz_y_lb) / 2.0, self._receptacle_width
+        ])
+        # Create gripper object such that it is at the middle of
+        # the screen.
+        data[self.robot] = np.array([(self.rz_x_lb + self.rz_x_ub) / 2.0,
+                                     (self.rz_y_lb + self.rz_y_ub) / 2.0,
+                                     self.gripper_width])
+        return State(data), goal_atoms
+
+    def _surface_xy_is_valid(self, x: float, y: float,
+                             existing_xys: Set[Tuple[float, float]]) -> bool:
+        # First, check that there isn't another screw at this location.
+        for (existing_x, existing_y) in existing_xys:
+            if (abs(existing_x - x) < self._screw_width
+                    and abs(existing_y - y) < self._screw_height):
+                return False
+        # Next, check that the location is valid.
```

Comment:
shouldn't this part be guaranteed by the bounds with which you're sampling? if not, change those bounds

Reply:
since we're using a normal distribution to sample, we can't enforce bounds

Reply:
ah right. FWIW, you could use a truncated normal, but this is fine too.